File: WEEK20/BFS_타겟 넘버.js
// BFS - 타겟 넘버

function solution(numbers, target) {
  var answer = 0;

  function depthFirst(idx, value) {
    if (idx === numbers.length) {
      if (value === target) {
        answer += 1;
      }
      return;
    }
    depthFirst(idx + 1, value + numbers[idx]);
    depthFirst(idx + 1, value - numbers[idx]);
  }

  depthFirst(0, 0);

  return answer;
}

console.log(solution([1, 1, 1, 1, 1], 3));

File: WEEK20/BFS_타겟 넘버.py
# BFS - 타겟 넘버

from collections import deque

def solution(numbers, target):
    answer = 0
    queue = deque([[0, numbers[0]], [0, -1 * numbers[0]]])

    while queue:
        (idx, value) = queue.popleft()
        idx += 1
        
        if idx < len(numbers):
            queue.append([idx, value + numbers[idx]])
            queue.append([idx, value - numbers[idx]])
        else:
            if value == target:
                answer += 1

    return answer